packages/lib-runtime/src/store/usePluginSubscription.ts

```diff
@@ -0,0 +1,94 @@
+import * as React from 'react';
+import type { PluginEventType, PluginConsumer } from '../types/store';
+import { useForceRender } from '../utils/react/useForceRender';
+import { usePluginConsumer } from './PluginStoreContext';
+
+/**
+ * React hook for subscribing to `PluginStore` events via the {@link PluginConsumer} interface.
+ *
+ * This hook implements the common {@link PluginConsumer} usage pattern:
+ *
+ * ```
+ * pluginConsumer.subscribe(() => {
+ *   // get current data from pluginConsumer
+ *   // compare current data with previous data
+ *   // re-render the component on data change
+ * }, eventTypes);
+ * ```
+ *
+ * Returns the current data as retrieved by `getData` function.
+ *
+ * Return value changes only if both of the following conditions are true:
+ * - `PluginStore` has emitted events of interest
+ * - current data has changed, compared to previously referenced data
+ *
+ * Therefore, the referential stability of the hook's result depends on a stable implementation
+ * of the `isSameData` function.
+ */
+export const usePluginSubscription = <TPluginData>(
+  getData: (pluginConsumer: PluginConsumer) => TPluginData,
+  isSameData: (prevData: TPluginData, nextData: TPluginData) => boolean,
+  eventTypes: PluginEventType[],
+): TPluginData => {
+  const pluginConsumer = usePluginConsumer();
+  const forceRender = useForceRender();
+
+  const getDataRef = React.useRef<typeof getData>(getData);
+  const isSameDataRef = React.useRef<typeof isSameData>(isSameData);
+  const eventTypesRef = React.useRef<typeof eventTypes>(eventTypes);
+
+  const hookResultRef = React.useRef<TPluginData>(getData(pluginConsumer));
+  const unsubscribeRef = React.useRef<VoidFunction>();
+  const isMountedRef = React.useRef(true);
+
+  const trySubscribe = React.useCallback(() => {
+    if (unsubscribeRef.current === undefined) {
+      unsubscribeRef.current = pluginConsumer.subscribe(() => {
+        const prevData = hookResultRef.current;
+        const nextData = getDataRef.current(pluginConsumer);
+        const dataChanged = !isSameDataRef.current(prevData, nextData);
+
+        if (dataChanged) {
+          hookResultRef.current = nextData;
+        }
+
+        if (isMountedRef.current && dataChanged) {
+          forceRender();
```

Comment:
Could this not be done with `useState`?
```
const [hookResult, setHookResult] = React.useState<TPluginData>(getData(pluginConsumer));
```

Reply:
Yes, I think so, good catch!

Reply:
Actually, doing it this way, we'd also need to put `hookResult` as a dependency of `trySubscribe` which would result in unnecessary recomputations of the memoized `trySubscribe` function.

Overall, I'd say `useRef` is great for infra-level hooks that want full control over the value and when to re-render, while `useState` is great for consuming app components that generally want re-render upon value change.

Reply:
Also, there'd be the issue when the `PluginStore` listener would fire but the component is no longer mounted. We still want to keep track of the current data, while not wanting to re-render since the component is not mounted. We can't have this level of control with `useState` hook.

Reply:
If you unsubscribed in an effect disposer function, how would this listener be invoked?
With `useState` you can pass a function to the `setter` function to get the current state thus removing it as a dependency in your function.

Reply:
Sounds good, we can use the so-called function updater form of `useState` value setter function, as described in [React docs](https://reactjs.org/docs/hooks-reference.html#usestate).